### DISCUSSION TITLE
change to semantics of query with no 'select' list

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#4744] [Design Proposals]
I just learned from the [migration guide](https://github.com/hibernate/hibernate-orm/blob/main/migration-guide.adoc#result-rows) that the semantics of a queries like:

    from Person p join p.address

has changed.

I did not know this, and I don't know what the reasoning was.

Perhaps there's a great reason, but at first sight this seems like a pretty bad idea to me:

1. A missing `select` clause has always been interpreted as approximately something like a SQL `select *`, i.e. return me everything in the `from` clause. This is something different. It returns me, what, only *root* entities in the `from` clause? Only the *first* root entity? I'm not sure.
2. It has the potential to break a huge number of existing HQL queries that have been working for decades, and the compiler won't help find them. Of all the things that are breaking in 6, this seems like pretty much the worst breakage of anything in terms of having a potentially wide impact on user code. Given that, the justification has got to be something a bit stronger than: "oh, well, we like this interpretation a bit better now".

So, what was behind this?

**Comment by sebersole** [comment]
The legacy behavior makes no sense.  

* HQL != SQL, and 
* no select != explicit select (*)

**Reply by gavinking** [reply]
> The legacy behavior makes no sense.

OK, that may be, but it makes sense _to me_, and so I would like you to explain why it doesn't make sense.

I mean, this is the sort of change we can only make if it's _clearly, objectively_ more correct.

I'm not seeing that. So help me see it.

**Reply by gavinking** [reply]
I mean, maybe just give me an example of a query where it doesn't make sense...

**Comment by gavinking** [comment]
So in trying to figure out exactly what the semantics of this are, I tried the following two very similar-looking HQL queries:

    from X x join Y y on x.id=y.id

And

    from X x, Y y where x.id=y.id

Clearly these are the "same" query, written using a slightly different syntax.

Now, in Hibernate 5, these two queries would have returned the same-shaped result list: a list of tuples of type `[X,Y]`.

But in Hibernate 6, the first query returns a List of `X`s, and the second query returns a list of tuples of shape `[X,Y]`.

I'm *really* struggling to see how the new behavior is better. In particular I don't see how it's the case that the legacy interpretation of the first query makes "no sense" whereas the legacy interpretation of the second query does make sense.

**Comment by sebersole** [comment]
> Clearly these are the "same" query, written using a slightly different syntax.

This is where we will disagree.  They are not the same

**Comment by sebersole** [comment]
Look, we simply disagree on this.  I'll let others chime in.  But I have no plans on reverting this

**Comment by gavinking** [comment]
This issue was fixed before H6 went out.

The resolution (suggested by @sebersole) was to use the result type to imply the `select` list.
